framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java

```diff
@@ -0,0 +1,492 @@
+            AnnotatedTypeMirror mr = combineTypeWithType(receiverType, methodReceiver);
+            mappings.put(methodReceiver, mr);
+        }
+
+        declMethodType =
+                (AnnotatedExecutableType) AnnotatedTypeReplacer.replace(declMethodType, mappings);
+
+        // Because we can't viewpoint adapt asMemberOf result, we adapt the declared method first, and sets the
+        // corresponding parts to asMemberOf result
+        methodType.setReturnType(declMethodType.getReturnType());
+        methodType.setReceiverType(declMethodType.getReceiverType());
+        methodType.setParameterTypes(declMethodType.getParameterTypes());
+        methodType.setTypeVariables(declMethodType.getTypeVariables());
+    }
+
+    protected boolean shouldAdaptMethod(ExecutableElement element) {
+        return !ElementUtils.isStatic(element);
+    }
+
+    @Override
+    public void viewpointAdaptTypeParameterBounds(
+            AnnotatedTypeMirror receiverType,
+            List<AnnotatedTypeParameterBounds> typeParameterBounds) {
+
+        List<AnnotatedTypeParameterBounds> adaptedTypeParameterBounds = new ArrayList<>();
+        for (AnnotatedTypeParameterBounds typeParameterBound : typeParameterBounds) {
+            AnnotatedTypeMirror adaptedUpper =
+                    combineTypeWithType(receiverType, typeParameterBound.getUpperBound());
+            AnnotatedTypeMirror adaptedLower =
+                    combineTypeWithType(receiverType, typeParameterBound.getLowerBound());
+            adaptedTypeParameterBounds.add(
+                    new AnnotatedTypeParameterBounds(adaptedUpper, adaptedLower));
+        }
+
+        typeParameterBounds.clear();
+        typeParameterBounds.addAll(adaptedTypeParameterBounds);
+    }
+
+    /**
+     * Viewpoint adapt declared type to receiver type, and return the result atm
+     *
+     * @param receiver receiver type
+     * @param declared declared type
+     * @return {@link AnnotatedTypeMirror} after viewpoint adaptation
+     */
+    protected AnnotatedTypeMirror combineTypeWithType(
+            AnnotatedTypeMirror receiver, AnnotatedTypeMirror declared) {
+        assert receiver != null && declared != null;
+
+        AnnotatedTypeMirror result = declared;
+
+        if (receiver.getKind() == TypeKind.TYPEVAR) {
+            receiver = ((AnnotatedTypeVariable) receiver).getUpperBound();
+        }
+        AnnotationMirror receiverAnnotation = extractAnnotationMirror(receiver);
+        if (receiverAnnotation != null) {
+            result = combineAnnotationWithType(receiverAnnotation, declared);
+            result = substituteTVars(receiver, result);
+        }
+
+        return result;
+    }
+
+    /**
+     * Extract modifier from {@link AnnotatedTypeMirror}.
+     *
+     * @param atm AnnotatedTypeMirror from which modifier is extracted
+     * @return modifier extracted
+     */
+    protected abstract AnnotationMirror extractAnnotationMirror(AnnotatedTypeMirror atm);
```

Comment:
could this have a default implementation at all in CF?

Reply:
Actually, no. Because each type system should provide the hierarchy in which viewpoint adaptation should happen, and this needs to be explicitly done.